File: Chapter03/src/designpatterns/decorator/java/DecoratorExample.java
package designpatterns.decorator.java;

public class DecoratorExample {
    public static void main(String[] args) {
        FrenchFries fries = new FrenchFries();

        Sandwich sandwich = new Sandwich();

        CheeseDecorator cheeseSandwich = new CheeseDecorator(sandwich);
        CheeseDecorator cheeseFrenchFries = new CheeseDecorator(fries);

        sandwich.cook();
        cheeseSandwich.cook();

        fries.cook();
        cheeseFrenchFries.cook();
    }
}

File: Chapter04/src/designpatterns/decorator/kotlin/DecoratorExample.kt
package designpatterns.decorator.kotlin

interface Recipe {
    fun cook()
}

class Sandwich : Recipe {
    override fun cook() {
        println("Preparing sandwich")
    }
}

class CheeseDecorator(val recipe: Recipe) : Recipe by recipe {
    override fun cook() {
        recipe.cook()
        garnish()
    }

    private fun garnish() {
        println("Keep calm and shred cheese...")
    }
}


fun main() {
    val sandwich = Sandwich().apply { cook() }
    val cheeseSandwich = CheeseDecorator(sandwich).apply { cook() }
}